src/server/models/userModel.ts

```diff
@@ -11,21 +12,21 @@ interface LoginDetails {
 }
 
 class UserModel {
-  private cachedLogins: {[id: string]: LoginDetails};
+  async getLoginDetails(login: string): Promise<LoginDetails|null> {
+    const tokensCollection = await firestore()
+        .collection('tokens');
 
-  constructor() {
-    // TODO: Move this to Firestore instead of in memory.
-    this.cachedLogins = {};
-  }
-
-  async getLoginDetails(username: string) {
-    for (const token of Object.keys(this.cachedLogins)) {
-      const details = this.cachedLogins[token];
-      if (details.username === username) {
-        return details;
-      }
+    const query = await tokensCollection.where('username', '==', login).get();
```

Comment:
FTR every document field is automatically indexed, so this should perform well: https://firebase.google.com/docs/firestore/query-data/indexing